### DISCUSSION TITLE
understanding how m controls memory size for GPU_IVF_PQ index

### DISCUSSION THREAD

**Discussion by akcopy** [milvus-io/milvus#30215] [Q&A and General discussion] [answered]
I don't quite understand how m controls memory size for a GPU_IVF_PQ index, does it control the dimensionality reduction? I read that the index is ~25% that of IVF_FLAT but isn't that purely due to the nbits ratio (32 bit float compressed to 8 bit)? How does m play a role here.

**Comment by yhmo** [comment, marked as answer]
The size of a IVF_PQ index is:
`(nb * m * nbits/8 + 2^nbits * d * 4 + nlist * d *4)`
 
nb is the number of vectors.
d is vector dimension.

Both m and nbits affect the size.

**Reply by akcopy** [reply]
what is the unit for this calculation? is it kB?

**Reply by yhmo** [reply]
yes, KB

**Reply by akcopy** [reply]
isn't this a bit strange? The first term would govern the size calculation for the most part, and it's basically nb * m * nbits / 8 (no reliance on d on the first term). With the default values (m=4, nbits = 8) we get nb * 4kB or 4kB/vector. But, with say a 512 dimension vector of 32 bit floats, it is only 2kB/vector. So the default IVF_PQ index consumes more memory than a flat index? Did I miss out something in my calculation?

**Reply by yhmo** [reply]
In Milvus, by default, a segment size is 512MB, each segment has an independent index.

- **If the dimension is 1024, d=1024**

Each segment has 130000+ vectors.  nb=130000
Typically, we set nlist=4 * sqrt(nb) and approximate to power of 2, nlist=4 * sqrt(130000)=1442=1024
Let's say m=4, nbits=8
First part = 130000 * 4 * 8/8 = 520KB
Second part = 2^8 * 1024 * 4 = 1MB
Third part = 1024 * 1024 * 4 = 4MB

In fact, we set m=4 for GPU_IVF_PQ because GPU memory capacity is small. 
For CPU IVF_PQ, we often set m = d/4 or d/8.
So, if we set m=1024/4=256.
First part = 130000 * 256 * 8/8 = 32MB
Second part = 2^8 * 1024 * 4 = 1MB
Third part = 1024 * 1024 * 4 = 4MB

If you set nbit=16, the size will be much larger(324MB).

- **If the dimension is 128, d=128**

Each segment has 1 million vectors.  nb=1000000
Typically, we set nlist=4 * sqrt(nb) and approximate to power of 2, nlist=4 * sqrt(1000000)=4000=4096
Let's say m=4, nbits=8

First part = 1000000 * 4 * 8/8 = 4MB
Second part = 2^8 * 128 * 4 = 128KB
Third part = 4096 * 128 * 4 = 2MB

In fact, we set m=4 for GPU_IVF_PQ because GPU memory capacity is small. 
For CPU IVF_PQ, we often set m = d/4 or d/8.
So, if we set m=128/4=32.
First part = 1000000 * 32 * 8/8 = 32MB
Second part = 2^8 * 1024 * 4 = 1MB
Third part = 1024 * 1024 * 4 = 4MB

If you set nbit=16, the size will be much larger(324MB).

**Reply by yhmo** [reply]
Summary: different parameters get different sizes.

**Reply by akcopy** [reply]
Thanks for the detailed explanation! I did the numbers again, this means that the equation you mentioned will give the number in terms of bytes, not kB. So for e.g. the 1st calculation: 130,000 * 4 * 8 / 8 = 520,000 (bytes) = 520KB. It is quite interesting that nbit has a much larger effect than m at higher dimension vector sizes. What I'm also interested to know is how does m affect the recall rates? Would higher m give more accurate results and what is the benefit of having a higher m value?

**Reply by yhmo** [reply]
I think a larger m gets a higher recall rate. You can read some papers to verify.
